Replace templating from bazel to Jinga2

roleroz opened the pull request:
This is needed to have a more descriptive platformio.ini template, so we can add sections only needed for some kinds of programming

roleroz committed:
Replace templating from bazel to Jinga2

This is needed to have a more descriptive platformio.ini template, so we
can add sections only needed for some kinds of programming

Kernald commented:
Do you have any examples that could be achieved with Jinja but not the built-in templates? Just my two cents, but adding an external dependency is not ideal.

roleroz commented:
Yes, this is based on a discussion that mum4k and I have been having about adding support for Arduino as ISP

Based on https://docs.platformio.org/en/latest/platforms/atmelavr.html there is a need for a handful of new lines in the platformio.ini file that shouldn't be in the platformio.ini file for direct programming. expand_template() doesn't allow for  conditionally adding full lines, so we have 3 options:
- Have multiple template files, this can easily get overwhelming, specially as we start supporting more programming methods (I'm currently working in Arduino as ISP, but OTA comes next in my list ;) )
- Create full lines in Starlark, and copy them over (ugly, as you have some part of your template in the template itself, and some in starlark)
- Use a more advanced templating engine (this solution, all templating happens in the template)

I understand that the extra dependency is not ideal (honestly it felt bad writing it in the documentation), but the cons of the other solutions weren't really that minor either

Kernald commented:
Thanks! The context is quite helpful. From what I've seen, other rulesets with such a need usually go for either 2), or a bash script instead of pulling an external dependency. That might be something to consider.

With that said, these rules already depend on PlatformIO, so sticking to Python is definitely the way to go if we're sticking to an external dependency here. One option to clean that up would be to look into rules_python which, I think (I'm not a Python developer so I'm not too familiar with that) have a way to load external dependencies from pip automatically. That might be worth looking into.

roleroz commented:
Bazel itself requires Python to run (I think), so that dependency is not that big

I agree with the Jinja dependency

I didn't know about the pip_* rules, those are awesome, and we should use them in the future. Having said that, their current documentation state "The package installation rules (pip_install, pip_parse etc.) are less stable. We may make breaking changes as they evolve.", so I'm hesitant to add them at this point. Will file an issue so we can track this and add it in the future

Kernald commented:
> Bazel itself requires Python to run (I think), so that dependency is not that big

I don't think it does, but PlatformIO does, so that one is a no-brainer anyway, we need it. The Python rules also have a way to specify a Python version (and I think automatically download an interpreter for that version, now) - which is quite useful to avoid being tied to the OS version.

> I agree with the Jinja dependency
> 
> I didn't know about the pip_* rules, those are awesome, and we should use them in the future. Having said that, their current documentation state "The package installation rules (pip_install, pip_parse etc.) are less stable. We may make breaking changes as they evolve.", so I'm hesitant to add them at this point. Will file an issue so we can track this and add it in the future

I didn't know they were still considered experimental. Our use-case would be extremely minimal though - no publishing or anything like that. It might be worth using those regardless of their experimental status.

roleroz committed:
Make Python use in template_renderer hermetic

Use rules_python (specifically pip_parse and friends) to manage a
hermetic python installation to run template_renderer, with this you
don't need to install jinja2 yourself

Remove printing from template_renderer, it was only used for debugging

roleroz committed:
Remove jinja2 from dependencies in documentation

Cleanup platformio/BUILD, there were unneded entries there

roleroz commented:
It took a while to get this right, but now bazel automatically manages Python versions and dependencies, making template rendering hermetic

roleroz committed:
Make template more readable

Found a way to manage whitespace that makes the template more readable

Kernald commented:
I think that's a much better user experience, thanks for taking the time to update this! It's definitely not in the scope of this PR, but I assume we can eventually do the same thing with platformio itself, which would be even better and not depending on the user environment at all.

roleroz commented:
That would be better, but yes, that will be another PR altogether :wink:

mum4k commented:
Thank you both for finding a better dependency management.

mum4k commented on platformio/template_renderer.py:
```diff
@@ -0,0 +1,24 @@
+import jinja2
```
Would you mind adding a top level file comment here explaining what this does and the context of its use for the posterity?

roleroz replied:
Done

mum4k commented on platformio/template_renderer.py:
```diff
@@ -0,0 +1,24 @@
+import json
+import sys
+
+# Read parameters
```
In case we ever need to expand this, can we add the usual python structure of functions instead of top level code here?

I.e. can we add a `def main` and then the usual:

```python
if __name__ == "__main__":
   main()
```

This will also give us an opportunity to document the main function, we could say something about the expected command line arguments.

roleroz replied:
Done

mum4k replied:
Great documentation, thank you.

roleroz committed:
Respond to code review comments

Add a main() function instead of top level code

Add documentation
